glyphs-reader/src/glyphdata.rs

```diff
@@ -599,9 +605,73 @@ impl GlyphData {
         Self::construct_category_via_agl(base_name)
     }
 
+    // https://github.com/googlefonts/glyphsLib/blob/c4db6b981d5/Lib/glyphsLib/glyphdata.py#L351
+    fn construct_production_name(&self, name: &str) -> Option<ProductionName> {
+        fn append_suffix(base_name: &mut String, suffix: Option<&str>) {
+            if let Some(suffix) = suffix {
+                base_name.push('.');
+                base_name.push_str(suffix);
+            }
+        }
+
+        fn is_u_name(name: &str) -> bool {
+            name.starts_with("u") && name[1..].bytes().all(|b| b.is_ascii_hexdigit())
+        }
+
+        let (base_name, suffix) = self.split_glyph_suffix(name);
+        let base_names = self
+            .split_ligature_glyph_name(base_name)
+            .unwrap_or_else(|| vec![base_name.into()]);
+        // Attempt to find a production name for each ligature component (or the whole base name).
+        // Return early if any such names have no GlyphData entry
+        // OR the entry doesn't specify a production name AND they aren't already AGLFN names...
+        let prod_names: Vec<SmolStr> = base_names
+            .into_iter()
+            .map(|name| {
+                self.query_no_synthesis(&name, None).and_then(|result| {
+                    result.production_name.map(Into::into).or_else(|| {
+                        // if no production name, return the name itself if already in AGLFN
+                        fontdrasil::agl::char_for_agl_name(name.as_ref()).map(|_| name)
+                    })
+                })
+            })
+            .collect::<Option<_>>()?;
+
+        // only (uniXXXX, uniYYYY, etc.) names with 4 hex digits can be concatenated using the
+        // more compact format uniXXXXYYYY... uXXXXX names for characters beyond BMP are joined
+        // in ligatures using the usual '_'.
+        let any_characters_outside_bmp = prod_names
+            .iter()
+            .any(|name| name.len() > 5 && is_u_name(name.as_ref()));
```

Comment:
nit: should be name.len() == 4, to my understanding you'd have to use uniXX if you for some reason decided to write a 2-hex name.

Reply:
It wasn't obvious to me that we'd validated that to be true, that we couldn't possibly get here with an invalid name. If we validated could we add a comment that notes as much? Or, better, have type that proves it because it can't exist unless valid. If we didn't validate then we need to handle malformed ones here.

Reply:
we don't validate production names at all. We assume the ones from glyphdata are 'valid' AGL names (they must be, that's the whole point of it), but users can set their own in the per-glyph properties. We handle malformed ones in so far as we strip invalid PS characters and prevent name collisions, that's all we care about here, or at least the original code we are trying to port over. This particular snippet of code that you commented on is simpy trying to see if there is any prod name that is in the form u + _more than_ 4 hex digits (beyond BMP) which signals we can't use the uniXXXX form for the compound ligature name.

Reply:
The glyphdata has to be valid. If users can specify them then we should check them and in some way handle it if they have the wrong number of hex digits, no?

Reply:
well, we currently don't validate them in glyphsLib/fontmake which this code is meant to match in functionality.. The worst that can happen is that we do as told and write a seeminly invalid production name but garbage in/garbage out?

Reply:
I don't want compilation to fail with a malformed production name, it should simply use that name as is. The whole AGL names reverse-unicode mapping is broken and works on a sort of best-effort basis

Reply:
but I'll add a test to confirm